In [40]:
import pandas as pd
import numpy as np
from datetime import datetime

In [41]:
raw_file_data = pd.read_csv('/content/cars_rawdata.csv')

In [42]:
raw_file_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

#`a and b) Handle absent values and remove units from the columns`

In [43]:
# Check for missing values in each column
absent_values = raw_file_data.isnull().sum()
print("absent Values are :")
print(absent_values)

absent Values are :
Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


#### **By above observation, Mileage, Engine, Power, Seats & New_Price has null or missing values.**

### **1. Handling Mileage column**

The error shows sample('19.67 kmpl') cannot convert to numeric.

**Solution**: Remove the units.

In [44]:
# Check absent values
print("Before Handling, No. of absent Values in Mileage Column:" ,  raw_file_data['Mileage'].isnull().sum())

# Regular Expression is used in str.extract() function to extract substrings
# \d+: Matches one or more digits.
# \.: Matches a literal dot (period).
# \d+: Matches one or more digits after the dot.
# astype(float) function is then used to convert the extracted substrings(str) to float.
raw_file_data['Mileage'] = raw_file_data['Mileage'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Mileage is reported in kmpl,
# it's reasonable to assume that missing values are likely to be missing at random.
# In such cases, filling with the median is a common strategy.
raw_file_data['Mileage'].fillna(raw_file_data['Mileage'].median(), inplace=True)

# Check absent values
print("After Handling, No. of absent Values in Mileage Column:" ,  raw_file_data['Mileage'].isnull().sum())


Before Handling, No. of absent Values in Mileage Column: 2
After Handling, No. of absent Values in Mileage Column: 0


### **2. Handling Engine column**

The error shows sample('300 CC') cannot convert to numeric.

**Solution**: Remove the units.

In [45]:
# Check absent values
print("After Handling, No. of absent Values in Engine Column:" ,  raw_file_data['Engine'].isnull().sum())
# Remove units
raw_file_data['Engine'] = raw_file_data['Engine'].str.extract('(\d+\.\d+|\d+)').astype(float)
# absent values for engine size are often handled by filling the mean value
raw_file_data['Engine'].fillna(raw_file_data['Engine'].mean(), inplace=True)
# Check absent values
print("Before Handling, No. of absent Values in Engine Column:" , raw_file_data['Engine'].isnull().sum())

After Handling, No. of absent Values in Engine Column: 36
Before Handling, No. of absent Values in Engine Column: 0


### **3. Handling Power Column**

The error shows sample('39.09 bhp') cannot convert to numeric.

**Solution**: Remove the units.

In [46]:
# Check absent values
print("After Handling, No. of absent Values in Power Column:" ,  raw_file_data['Power'].isnull().sum())
# Remove units
raw_file_data['Power'] = raw_file_data['Power'].str.extract('(\d+\.\d+|\d+)').astype(float)
# Missing values for Power are often handled by filling the mean value
raw_file_data['Power'].fillna(raw_file_data['Power'].mean(), inplace=True)
# Check absent values
print("Before Handling, No. of absent Values in Power Column:" , raw_file_data['Power'].isnull().sum())

After Handling, No. of absent Values in Power Column: 36
Before Handling, No. of absent Values in Power Column: 0


### **4. Handling Seats Column**

In [47]:
# Check absent values
print("Before Handling, No of absent Values in Seats Column:" ,  raw_file_data['Seats'].isnull().sum())

# absent values for seats of the car are often handled by filling the mean value
raw_file_data['Seats'].fillna(raw_file_data['Seats'].mean(), inplace=True)
# Check absent values
print("After Handling, No. of absent Values in Seats Column:" ,  raw_file_data['Seats'].isnull().sum())

Before Handling, No of absent Values in Seats Column: 38
After Handling, No. of absent Values in Seats Column: 0


### **5. Handling New_Price Column**

In [48]:
# Drop the New_Price column since it has a high number of missing values
raw_file_data.drop('New_Price', axis=1, inplace=True)

**View the no null values handled DataFrame.**

In [ ]:
absent_values = raw_file_data.isnull().sum()
absent_values

Unnamed: 0           0
Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage              0
Engine               0
Power                0
Seats                0
Price                0
dtype: int64

# `c)  Change the 'Fuel_Type' and 'Transmission' coulmns into numerical one hot encoded value`

In [49]:
# Encoding Categorical Variables by using get_dummies function from pandas
raw_file_data = pd.get_dummies(raw_file_data, columns=['Fuel_Type', 'Transmission'])

# Display the modified dataset
print(raw_file_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

# `d) create new feature and add this column to dataset  `

### **Creating car_age feature which define the age of the car.**

In [50]:
# Calculate present year
this_year = datetime.now().year

# Subtract "Year" column from the current year to calculate the age of the car
raw_file_data['car_age'] = this_year - raw_file_data['Year']

# Display the modified dataset with the new feature
print(raw_file_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

#### **Add Mileage_Category feature to define mileage value by category.**

In [51]:
# Define the mileage categories such as mileage bins and mileage lables
mileage_bins = [0, 10, 20, 30, float('inf')]
mileage_labels = ['Very Low', 'Low', 'Medium', 'High']

# Create a new feature 'Mileage_Category' based on the mileage groups
raw_file_data['Mileage_Category'] = pd.cut(raw_file_data['Mileage'], bins=mileage_bins, labels=mileage_labels, right=False)

# Display the modified dataset with the new feature
print(raw_file_data.head())


   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

## e) Apply the procedures select, filter, rename, mutate, arrange, and summarize with group by (or the corresponding Python actions) to this dataset.


In [53]:
# Selecting required columns
selected_data = raw_file_data[['Name', 'Location', 'Year', 'Mileage', 'Price', 'car_age']]
# Display the selected data
print("Selected Data:")
print(selected_data.head())

Selected Data:
                               Name    Location  Year  Mileage  Price  car_age
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015    19.67  12.50        9
1                      Honda Jazz V     Chennai  2011    13.00   4.50       13
2                 Maruti Ertiga VDI     Chennai  2012    20.77   6.00       12
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013    15.20  17.74       11
4            Nissan Micra Diesel XV      Jaipur  2013    23.08   3.50       11


In [54]:
# Filtering data on yearly basis
filtered_data = raw_file_data[raw_file_data['Year'] >= 2017]
print("Filtered Data Year greater than or equal to 2019):")
print(filtered_data.head())

Filtered Data Year greater than or equal to 2019):
    Unnamed: 0                                     Name Location  Year  \
8           10                         Maruti Ciaz Zeta    Kochi  2018   
15          17                    Maruti Swift DDiS VDI   Jaipur  2017   
26          28                      Honda WRV i-VTEC VX    Kochi  2018   
34          36            Maruti Vitara Brezza ZDi Plus    Kochi  2018   
35          37  Mahindra KUV 100 mFALCON G80 K6 5str AW    Delhi  2017   

    Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
8               25692      First    21.56  1462.0  103.25    5.0   9.95   
15              25000      First    28.40  1248.0   74.00    5.0   5.99   
26              37430      First    17.50  1199.0   88.70    5.0   9.90   
34              50075      First    24.30  1248.0   88.50    5.0   8.63   
35              52000      First    18.15  1198.0   82.00    5.0   4.85   

    Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_P

In [56]:
#  changing the column name
renamed_data = raw_file_data.rename(columns={'Engine': 'Engine_Displacement', 'Power': 'Horsepower'})

print("\n Renamed Columns : ")
print(renamed_data.head())


 Renamed Columns : 
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine_Displacement  Horsepower  \
0              41000      First    19.67               1582.0      126.20   
1              46000      First    13.00               1199.0       88.70   
2              87000      First    20.77               1248.0       88.76   
3              40670     Second    15.20               1968.0      140.80   
4              86999      First    23.08               1461.0       63.10   

   Seats  Price  Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0    5.0  12.50           

In [57]:
# (creating new columns)
raw_file_data['Price_Category'] = pd.cut(raw_file_data['Price'], bins=[0, 5, 10, 15, 20, 25, 30], labels=['<5', '5-10', '10-15', '15-20', '20-25', '25-30'])

print("\nMutated Data (Added Price_Category column):")
print(raw_file_data.head())


Mutated Data (Added Price_Category column):
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1               

In [58]:
# Arranging data on yearly basis in desc order
arranged_data = raw_file_data.sort_values(by='Year', ascending=False)

print("\nArranged Data (Sorted by Year in descending order):")
print(arranged_data.head())


Arranged Data (Sorted by Year in descending order):
      Unnamed: 0                             Name    Location  Year  \
5405        5560        Renault KWID RXT Optional       Kochi  2019   
942          975  Ford Endeavour 2.2 Trend AT 4X2       Kochi  2019   
5533        5690        Maruti Omni 5 Seater BSIV  Coimbatore  2019   
770          796            Mahindra XUV500 W9 AT  Coimbatore  2019   
4267        4399       Maruti Swift Dzire AMT ZDI     Chennai  2019   

      Kilometers_Driven Owner_Type  Mileage  Engine  Power  Seats  Price  \
5405               6568      First    25.17   799.0   53.3    5.0   5.09   
942               11209      First    12.62  2198.0  158.0    7.0  31.15   
5533               4721      First    14.00   796.0   35.0    5.0   4.11   
770               19654      First    14.00  2179.0  155.0    7.0  17.63   
4267              65000      First    26.59  1248.0   74.0    5.0   6.75   

      Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \

In [59]:
# validating data with group by operations
summary_data = raw_file_data.groupby('Location')['Price'].mean()

print("\nSummary Data (Average Price by Location):")
print(summary_data)


Summary Data (Average Price by Location):
Location
Ahmedabad      8.567248
Bangalore     13.482670
Chennai        7.958340
Coimbatore    15.160206
Delhi          9.881944
Hyderabad      9.997423
Jaipur         5.916725
Kochi         11.309109
Kolkata        5.733924
Mumbai         9.592546
Pune           6.951000
Name: Price, dtype: float64


# `Save raw_file_data(DataFrame) in New CSV`

In [60]:
raw_file_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 0                   0                 1   
2                 1                   0                 0   
3                 1                   0                 0   
4                 1                   0                 0   

   Transmission_Automatic  Transmission_Manual  car_age Mileage_Category  \
0                       0                    1        9              Low   
1                       0                    1       13              Low   
2                       0                    1       12           Medium   
3                       1                    0       11              Low   
4                       0                    1       11           Medium   

  Price_Category  
0          10-15  
1             <5  
2           5-10  
3          15-20  
4             <5

In [61]:
#Save the data into clean data csv
raw_file_data.to_csv("/content/clean_data.csv")